### PR TITLE
ENH: inconsistent naming convention for read_excel column selection (#4988)

### LINKED ISSUE #4988: ENH: inconsistent naming convention for read_csv and read_excel column selection
`read_csv` uses the `usecols` parameter, and the `read_excel` function uses `parse_cols` parameter. Both parameter do basically the same thing:

```
usecols : array-like
    Return a subset of the columns.
    Results in much faster parsing time and lower memory usage.
```

```
parse_cols : int or list, default None
    * If None then parse all columns,
    * If int then indicates last column to be parsed
    * If list of ints then indicates list of column numbers to be parsed
    * If string then indicates comma separated list of column names and
      column ranges (e.g. "A:E" or "A,C,E:F")
```

Also maybe supporting the same thing that `parse_cols` does in `usecol` wouldn't be so bad?

### PR DESCRIPTION
**Pull request by abarber4gh**
- update API to use ‘usecols’ instead of ‘parse_cols’.  still functionally the same as ‘parse_col’,
- added test cases for ‘usecols’, added assert_produces_warning(FutureWarning) to other test
cases that use ‘parse_cols’.
- refactor column use column parsing to only occur once per sheet.
- updated whats new with deprecated parse_col argument.

 - [X] closes #4988 
 - [X] tests added / passed
 - [X] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [X] whatsnew entry
 - [X] packers.* BENCHMARKS NOT SIGNIFICANTLY CHANGED.

### CODE REVIEW

**Comment by chris-b1** [comment]
FYI, just discussing a bug related to usecols #15316, might be easiest to work that change into your refactoring, or can be separate, cc @stanleyguan

File: doc/source/whatsnew/v0.21.0.txt [hunk 1]
```diff
@@ -35,6 +35,7 @@ Other Enhancements
 - ``RangeIndex.append`` now returns a ``RangeIndex`` object when possible (:issue:`16212`)
 - :func:`to_pickle` has gained a protocol parameter (:issue:`16252`). By default, this parameter is set to `HIGHEST_PROTOCOL <https://docs.python.org/3/library/pickle.html#data-stream-format>`__
 - :func:`api.types.infer_dtype` now infers decimals. (:issue: `15690`)
+- :func:`read_excel` now allows a column character list (E.G. ['A', 'C', 'D']) with the ``usecols`` parameter (:issue:`4988`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: we are listed this in deprecations so this is unecessary

File: doc/source/whatsnew/v0.21.0.txt [hunk 2]
```diff
@@ -60,6 +61,7 @@ Other API Changes
 Deprecations
 ~~~~~~~~~~~~
 - :func:`read_excel()` has deprecated ``sheetname`` in favor of ``sheet_name`` for consistency with to_excel() (:issue:`10559`).
+- :func:`read_excel()` has deprecated ``parse_cols`` in favor of ``usecols`` for consistency with other read_ functions (:issue:`4988`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.21.0.txt]: ``pd.read_*`` functions

File: pandas/io/excel.py [hunk 3]
```diff
@@ -111,12 +111,16 @@
...
+usecols : int or list, default None
     * If None then parse all columns,
-    * If int then indicates last column to be parsed
-    * If list of ints then indicates list of column numbers to be parsed
+    * If int then indicates last column to be used
+    * If list of ints then indicates list of column numbers to be used
     * If string then indicates comma separated list of column names and
       column ranges (e.g. "A:E" or "A,C,E:F")
+parse_cols : int or list, default None
+    .. deprecated:: 0.21.0
```
Comment (jreback) [inline pandas/io/excel.py]: hmm, didn't know sphinx had this deprecated tag. @TomAugspurger @jorisvandenbossche should maybe change other deprecations to use this if it looks nice.

File: pandas/tests/io/test_excel.py [hunk 4]
```diff
@@ -169,54 +169,155 @@ def test_parse_cols_int(self):
...
         # TODO add index to xls file)
         tm.assert_frame_equal(df1, dfref, check_names=False)
         tm.assert_frame_equal(df2, dfref, check_names=False)
 
+    def test_usecols_int(self):
+        # GH4988: inconsistent naming convention for read_excel column select
+        dfref = self.get_csv_refdf('test1')
+        dfref = dfref.reindex(columns=['A', 'B', 'C'])
+        df1 = self.get_exceldf('test1', 'Sheet1', index_col=0, usecols=3)
+        df2 = self.get_exceldf('test1', 'Sheet2', skiprows=[1], index_col=0,
```
Comment (jreback) [inline pandas/tests/io/test_excel.py]: so you need to change ALL tests to use the new one (usecols), except for a single test to actually hit the deprecation.

File: pandas/tests/io/test_excel.py [hunk 5]
```diff
@@ -169,54 +169,155 @@ def test_parse_cols_int(self):
...
+        dfref = dfref.reindex(columns=['A', 'B', 'C'])
+        df1 = self.get_exceldf('test1', 'Sheet1', index_col=0,
+                               usecols=['A', 'B', 'C', 'D'])
+        with tm.assert_produces_warning(FutureWarning, check_stacklevel=False):
+            df2 = self.get_exceldf('test1', 'Sheet1', index_col=0,
+                                   parse_cols=3)
+
+        tm.assert_frame_equal(df1, dfref, check_names=False)
+        tm.assert_frame_equal(df2, dfref, check_names=False)  # backward compat
+
```
Comment (jreback) [inline pandas/tests/io/test_excel.py]: are these new tests?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/excel.py (hunk 3 above)
Comment (abarber4gh) [inline pandas/io/excel.py]: appears as:
> **parse_cols** _: int or list, default None_
> _Deprecated since version 0.21.0:_ Use _usecols_ instead

Comment (TomAugspurger) [inline pandas/io/excel.py]: http://pandas-docs.github.io/pandas-docs-travis/generated/pandas.read_excel.html Has an example from sheetname. Looks ok

Comment (jreback) [inline pandas/io/excel.py]: @TomAugspurger if we like this (and lgtm), let's make an issue to updated all DEPRECATED in codebase with this directive

File: pandas/io/excel.py [hunk 6]
```diff
@@ -312,11 +327,6 @@ def _range2cols(areas):
             >>> _range2cols('A,C,Z:AB')
             [0, 2, 25, 26, 27]
             """
-            def _excel2num(x):
-                "Convert Excel column name like 'AB' to 0-based column index"
-                return reduce(lambda s, a: s * 26 + ord(a) - ord('A') + 1,
```
Comment (jreback) [inline pandas/io/excel.py]: is there a reason you are changing this code? does this involve the deprecation somehow? if you are cleaning/fixing, pls do in another PR.

File: pandas/tests/io/test_excel.py [hunk 7]
```diff
@@ -165,58 +165,100 @@ def setup_method(self, method):
...
+                               usecols=[0, 2, 3])
+        with tm.assert_produces_warning(FutureWarning, check_stacklevel=False):
+            df3 = self.get_exceldf('test1', 'Sheet1', index_col=0,
+                                   parse_cols=[0, 2, 3])
+
         tm.assert_frame_equal(df1, dfref, check_names=False)
         tm.assert_frame_equal(df2, dfref, check_names=False)
+        tm.assert_frame_equal(df3, dfref, check_names=False)  # backward compat
 
-    def test_parse_cols_str(self):
```
Comment (jreback) [inline pandas/tests/io/test_excel.py]: leave the original tests structure (sure you can change the name to conform), but don't change the tests (in THIS PR).

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/excel.py (hunk 6 above)
Comment (abarber4gh) [inline pandas/io/excel.py]: code changed to allow passing list of strings, similar to other `read_*` functions, and keep the `read_excel` "superpower" to pass a string that is parsed as mentioned in the original issue (#4988, point 3).

File: pandas/tests/io/test_excel.py (hunk 5 above)
Comment (abarber4gh) [inline pandas/tests/io/test_excel.py]: yes, new tests for the new functionality.

File: pandas/io/excel.py (hunk 6 above)
Comment (jreback) [inline pandas/io/excel.py]: and that's fine, but needs to be in a separate PR from the deprecation change.

Comment (abarber4gh) [inline pandas/io/excel.py]: ok, opened #16510 for this functionality.  will re-submit with only the argument change (parse_cols -> usecols).

**Comment by TomAugspurger** [comment]
@abarber4gh can you rebase, now that the excel tests are running?

File: pandas/tests/io/test_excel.py (hunk 7 above)
Comment (abarber4gh) [inline pandas/tests/io/test_excel.py]: tests are back to original but with changed function & kwarg names.

**Comment by jreback** [comment]
can you rebase / update according to comments

**Comment by jreback** [comment]
closing as stale. pls ping to reopen if you want to continue.

**Pull request closed without merging**